cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+              (iii) If there are gates remaining in the current timesteps, consider a set of
+                    candidate swaps on them and rank them based on a heuristic cost function. Pick
+                    the swap that minimises the cost and use it to update our logical to physical
+                    mapping. Repeat from (i).
+    """
+
+    def __init__(self, device_graph: nx.Graph):
+        """Initializes the circuit routing transformer."""
+
+        if nx.is_directed(device_graph):
+            raise ValueError("Device graph must be undirected.")
+        self.device_graph = device_graph
+
+    def __call__(
+        self,
+        circuit: 'cirq.AbstractCircuit',
+        *,
+        max_search_radius: int = 8,
+        preserve_moment_structure: bool = False,
+        tag_inserted_swaps: bool = False,
+        initial_mapper: Optional['cirq.AbstractInitialMapper'] = None,
+        context: Optional['cirq.TransformerContext'] = None,
+    ) -> 'cirq.AbstractCircuit':
+        """Transforms the given circuit to make it executable on the device.
+
+        Since routing doesn't necessarily modify any specific operation and only adds swaps
+        before /after operations to ensure the circuit can be executed, tagging operations with
+        tags from context.tags_to_ignore will have no impact on the routing procedure.
+
+        Args:
+            circuit: the input circuit to be transformed.
+            max_search_radius: the maximum number of times the cost function can be iterated for
+                convergence.
+            preverse_moment_structure: whether or not the transfomer should preserve the given
+                moment structure of 'circuit'.
+            tag_inserted_swaps: whether or not a RoutingSwapTag should be attched to inserted swap
+                operations.
+            initial_mapper: an initial mapping strategy (placement) of logical qubits in the
+                circuit onto physical qubits on the device.
+            context: transformer context storing common configurable options for transformers.
+
+        Returns:
+            The routed circuit executable on the harware with the same unitary as 'circuit'.
+
+        Raises:
+            ValueError: if circuit has operations that act on 3 or more qubits.
+        """
+
+        # 0. Handle CircuitOperations by unrolling them.
+        if context is not None and context.deep is True:
+            circuit = transformer_primitives.unroll_circuit_op(circuit, deep=True)
+        if not all(protocols.num_qubits(op) <= 2 for op in circuit.all_operations()):
+            raise ValueError("Input circuit must only have ops that act on 1 or 2 qubits.")
+
+        # 1. Do the initial mapping of logical to physical qubits.
+        if initial_mapper is None:
+            initial_mapper = line_initial_mapper.LineInitialMapper(self.device_graph)
+        self._initial_mapping = initial_mapper.initial_mapping(circuit)
```

Comment:
it's pretty bad form to declare/define member attributes outside of `__init__`. Does the linter not yell at you about this?

When defining classes, it's always better to start from thinking of them as a bundle of data rather than a bundle of functions. What's the data that defines this transformer?

In the same vein: I'd suggest trying to work some or many of the helper methods into free functions. It's a lot lot easier to reason about functions that take inputs and return outputs rather than methods that mutate a (semi-)global state

Reply:
I think making the initial mapping and mapping manager passable objects to functions makes sense. I'll make those changes and see which functions don't need to know about `_cost()` and can be made free.